### ISSUE TITLE
H264 encoding - Criware/Scaleform SDK request

### ISSUE THREAD

**Issue by emoose** [emoose/NieRAutomata-LodMod#8]
With LodMod v0.77 we've been able to patch in support to the game for H264, by using H264-USMs from other titles it's been confirmed that it seems to work fine.

Unfortunately we don't actually have any method to create our own H264 USM though... the only tool that can even create normal MPEG1-USMs is some ancient version of Scaleform VideoEncoder, which has it's own issues (eg. out-of-memory issues when encoding 4k videos)

Apparently that VideoEncoder was updated however, and also added a **medianocheH264.exe** file for encoding H264-USM files, but sadly doesn't look like that update was ever publicly posted anywhere...

The Criware SDK package also seems to include encoders with it (**sofdec2encH264.exe**...) but again it doesn't seem that this has ever been released publicly neither (only ones that have been are old ones without the Sofdec2 tools included ;_;)

TL;DR: If anyone ever had access to Scaleform GFx or Criware SDKs, could you check if you have any version of **medianocheH264.exe** or **sofdec2encH264.exe**? If you're able to share it I'd be extremely grateful, would help NA cutscene modders out a lot!

(My email is on my github profile page if you want to get in contact: https://github.com/emoose)

**Comment by emoose** [comment]
Happy to say this request was answered, many thanks to the person who contacted me!

**Comment by ThRequiemOwl** [comment]
hey chum, could you please share with me the encoder please?

**Comment by BigheadSMZ** [comment]
Was the encoder ever uploaded or shared anywhere? There really isn't much point to having H.264 support without a means to actually creating it. I've been interested in running some tests to see what kind of results I can get, but I have no idea of a tool that encode H264 into USM files. The only clue is an executable with the name "medianoche" which comes with Scaleform Video Encoder, but public versions are old and do not support H264 codec. Any leads would be greatly appreciated.

**Comment by emoose** [comment]
The v4.6 encoder at https://www.nexusmods.com/witcher3/mods/3505 should work with H264, there were also some recent criware SDKs released that include H264 too, can't find link to them atm but they were on archive.org iirc.

**Comment by BigheadSMZ** [comment]
Thanks for reply, it's appreciated. Ironically not long after I posted, I found a working version here too:
https://modding.wiki/en/aewff/developers/usmconversion

I managed to encode the first Nier: Automata USM video (ev0010) with H.264 and it worked in-game, now I'm working on nailing down the workflow writing some automation scripts so it's easier to batch them. Rather than use the video from the base game, I tried using one from the 1080p versions of the packs here: https://www.nexusmods.com/nierautomata/mods/37

Interestingly enough, the audio did not go out of sync like the 4k H.264 video did. I'm not sure if it's the way they were encoded or if 4K is unnoticeably slower. I was surprised to see that audio is not embedded in the files. Either way, I plan on running some experiments now to see what kind of results I can get. Currently the 4K pack there is the only one offering H.264 encodes and it appears to have issues, so I figured I would give it a go to see what I can come up with.

**Comment by BigheadSMZ** [comment]
@emoose : Sorry to bother you again and sorry if this is not the place, I'm not sure where else to contact you. It's been a long time since I've tried to encode videos outside of just tossing a video into my handbrake script to cut down on file size, so I'm a bit out of touch with some of the tools and processes that are out there. Despite that, I've made some good progress since this morning. I  found many useful tools, though opted for command line versions and put together a workflow that can go from original USM to reencoded USM. The process looks a bit something like this:

1.) Original USM to M2V via USMToolkit
2.) M2V to raw AVI via ffmpeg
3.) AVI to upscaled H.264 MP4 via video2x
4.) Upscaled MP4 to raw AVI via ffmpeg
5.) Upscaled raw AVI to H.264 USM via medianocheH264

No problems when working with 1080p videos. But unfortunately, medianocheH264 is where the process breaks when trying to process 2160p videos. It just flat out gives an error "Unsupported video". I was thinking maybe I'm encoding something wrong along the way, but exact settings work from start to finish when working with 1080p videos. Is the problem medianoche doesn't support 4k videos? I haven't taken a look at the archive.org links yet, but I'm going to do that now.

**Comment by emoose** [comment]
The medianoche in the sf_video_4.6.33_windows_msvc11_lib_4K-ready release should work with 4k, had to patch some stuff specifically to remove some resolution limits in there, which ended up being used for the other 4K mod.

That medianoche version is pretty old and might be missing some features in the newer ones though, I think the creator of the 4K mod actually ended up encoding using normal x264.exe first and then just passed the encoded result to medianoche for muxing into usm, sadly can't remember all the details of it now though.

Someone else was asking me the same a couple months ago (and I still only had the same vague answers), but they did mention they got in contact with the mod creator and was able to find how they made it, maybe worth contacting them?

**Comment by emoose** [comment]
Looking through some DMs he did mention the newer tools were working for him too:

> I somewhat recently found much newer CRIWARE tools and I can happily report that 4K H.264 USMs created with those work perfectly without bitrate limitation with LodMod in NA - so there was nothing wrong with the LodMod fix. Although I only use the tools to mux H.264 video that I make with regular x264 (can do CRF encode and such) via commandline with the tools - it can mux that to USM without any reencoding. Anyways that's just something I figured you wouldn't mind knowing in case he didn't already tell you.

The newer tools he mentioned were the new-criware-sdk files, not really sure what the commands for muxing are though.

**Comment by BigheadSMZ** [comment]
Interesting, I had the same theory about "crispeee_x264.dll" that it could just be launched directly. I found a log file (medianocheH264.log) that showed that medianoche was just a frontend sending commands. When encoding 1080p movies, I was a bit annoyed that it automatically sets the thread priority to "Below Normal" so I started poking around trying to figure out how to override that. The thought then hit that I could just send the commands to the DLL, but well, can just change priority from Task Manager after starting so I was like meh. But now it's worth looking deeper into this. Thanks for the tip, I'm not defeated yet.

**Comment by BigheadSMZ** [comment]
So the issue I've been facing this whole time is the crack for "sofdec2enc.exe" not working. I extrapolated the commands that were sent to crispeee and softdec2enc and managed to just pass them directly. When getting to softdec2, again I was faced with "license verification failed". This has been the only thing holding me up. I think I may have gotten it to work now, I uninstalled everything, redownloaded, reinstalled, and reapplied the crack. I am now able to use the sofdec2 wizard and a video is supposedly being encoded into USM.

**Comment by BigheadSMZ** [comment]
So I already filled this page with far too many posts, but I wanted to say that I am finally able to create movies at any resolution using H.264. One thing that was really bothering me is the speed of the tools that come with the CriWare stuff, in particular, **crispeee_x264.dll**. Looking at the commands made me realize that it's nothing more than an offshoot of x264 which can be found [here](https://www.videolan.org/developers/x264.html). So I figured, lets try that. And it works fine and is literally thousands of times faster. So the process ends up being this:

- extract USM to M2V
- encode to lossless AVI with ffmpeg.exe
- 2 pass encode with x264.exe
- the pass with crimed.exe (still not sure if this is even necessary)
- the final pass with sofdec2.exe to USM

Doing it this way encodes the 157 frame video in less than 20 seconds, as opposed to nearly 30 minutes it was taking. I'm not sure why crispeee is so slow, one guess is it's not even trying to make full use of my hardware. So that's it, it's now possible to actually pull this off. I'm not sure what kind of upscaler to use (currently just trying out video2x), and what settings I'll ultimately settle on, but it's nice to see that this process is actually possible. I'll log here all tools that make this possible.

**USMToolKit:** https://github.com/Rikux3/UsmToolkit
**ffmpeg:** https://ffmpeg.org/download.html
**x264 Encoder:** https://www.videolan.org/developers/x264.html
**Sofdec2:** https://archive.org/details/new-criware-sdk

And of course, video upscaler of the user's choice. Sorry for all the posts, and thank you for all the help.

**Comment by BigheadSMZ** [comment]
Just wanted to say that I finally finished and published my mod on NexusMods.
https://www.nexusmods.com/nierautomata/mods/688/?tab=description

Thanks for all the help!

**Comment by FINALZ-27980** [comment]
Scaleform SDK Download
https://github.com/FINALZ-27980/scaleform_sdk/releases
Is 4.3.28 Version
